packages/dds/migration-shim/src/shimDeltaConnection.ts

```diff
@@ -23,47 +23,48 @@ import { type IShimDeltaHandler } from "./types";
  * modifications on the current ChannelDeltaConnection.
  */
 export class ShimDeltaConnection implements IDeltaConnection {
-	public constructor(private readonly deltaConnection: IDeltaConnection) {}
-	private _handler: IShimDeltaHandler | undefined;
-	private get handler(): IShimDeltaHandler {
-		assert(this._handler !== undefined, "Missing delta handler");
-		return this._handler;
-	}
+	public constructor(
+		private readonly deltaConnection: IDeltaConnection,
+		private readonly shimDeltaHandler: IShimDeltaHandler,
+	) {}
+	private isShimDeltaHandlerAttachedToConnection = false;
 
 	public get connected(): boolean {
 		return this.deltaConnection.connected;
 	}
 
 	// Should we be adding some metadata here?
 	public submit(messageContent: unknown, localOpMetadata: unknown): void {
-		// The expectation is to currently add some form of V2 stamp to the metadata here that can eventually be
-		// ignored by a SharedObject V2
-		// TODO: stamp messageContent with V2 metadata
+		// TODO: stamp messageContent with V2 metadata - this is not the final implementation
+		this.shimDeltaHandler.stampV2Ops(messageContent);
 		this.deltaConnection.submit(messageContent, localOpMetadata);
 	}
 
-	// Note we have an option of also adding submit to the IShimDeltaHandler interface
-	// That way in the submit method above we could call this.handler.preSubmit(messageContent, localOpMetadata) to
-	// stamp the v2 ops properly.
-	public preAttach(handler: IShimDeltaHandler): void {
-		assert(this._handler === undefined, "Should only attempt to preAttach once!");
-		this._handler = handler;
-		this.deltaConnection.attach(handler);
+	// The underlying ChannelDeltaConnection can only attach one IDeltaHandler.
+	// Maybe we could do this in the constructor?
+	private attachShimDeltaHandler(): void {
+		assert(this.shimDeltaHandler.hasTreeDeltaHandler(), "No tree handler to process op");
+		this.deltaConnection.attach(this.shimDeltaHandler);
+		this.isShimDeltaHandlerAttachedToConnection = true;
 	}
```

Comment:
I'd probably inline this in the `attach()` method now, since it's the only callsite.  This also lets you drop the assert since you explicitly attachTreeDeltaHandler in the same synchronous block.